index.ts

```diff
@@ -0,0 +1,476 @@
+	zMove?: string | true | null;
+	zMoveFrom?: string;
+	zMoveType?: TypeName;
+	zMoveUser?: string[] | null;
+}
+
+class Items {
+	static forGen(gen: Generation): Promise<Items>;
+
+	get(name: string): Item | undefined;
+	getByID(id: ID): Item | undefined;
+	next(): [ID, Item] | true;
+}
+
+// moves.ts
+
+type MoveFlag = 1 | 0;
+
+interface MoveFlags {
+	authentic?: MoveFlag; // Ignores a target's substitute.
+	bite?: MoveFlag; // Power is multiplied by 1.5 when used by a Pokemon with the Ability Strong Jaw.
+	bullet?: MoveFlag; // Has no effect on Pokemon with the Ability Bulletproof.
+	charge?: MoveFlag; // The user is unable to make a move between turns.
+	contact?: MoveFlag; // Makes contact.
+	dance?: MoveFlag; // When used by a Pokemon, other Pokemon with the Ability Dancer can attempt to execute the same move.
+	defrost?: MoveFlag; // Thaws the user if executed successfully while the user is frozen.
+	distance?: MoveFlag; // Can target a Pokemon positioned anywhere in a Triple Battle.
+	gravity?: MoveFlag; // Prevented from being executed or selected during Gravity's effect.
+	heal?: MoveFlag; // Prevented from being executed or selected during Heal Block's effect.
+	mirror?: MoveFlag; // Can be copied by Mirror Move.
+	mystery?: MoveFlag; // Unknown effect.
+	nonsky?: MoveFlag; // Prevented from being executed or selected in a Sky Battle.
+	powder?: MoveFlag; // Has no effect on Pokemon which are Grass-type, have the Ability Overcoat, or hold Safety Goggles.
+	protect?: MoveFlag; // Blocked by Detect, Protect, Spiky Shield, and if not a Status move, King's Shield.
+	pulse?: MoveFlag; // Power is multiplied by 1.5 when used by a Pokemon with the Ability Mega Launcher.
+	punch?: MoveFlag; // Power is multiplied by 1.2 when used by a Pokemon with the Ability Iron Fist.
+	recharge?: MoveFlag; // If this move is successful, the user must recharge on the following turn and cannot make a move.
+	reflectable?: MoveFlag; // Bounced back to the original user by Magic Coat or the Ability Magic Bounce.
+	snatch?: MoveFlag; // Can be stolen from the original user and instead used by another Pokemon using Snatch.
+	sound?: MoveFlag; // Has no effect on Pokemon with the Ability Soundproof.
+}
+
+type MoveCategory: 'Physical' | 'Special' | 'Status';
+
+type MoveTarget =
+	// single-target
+	'normal'|'any'|'adjacentAlly'|'adjacentFoe'|'adjacentAllyOrSelf'|
+	// single-target, automatic
+	'self'|'randomNormal'|
+	// spread
+	'allAdjacent'|'allAdjacentFoes'|
+	// side and field
+	'allySide'|'foeSide'|'all';
+
+interface SelfEffect {
+	volatileStatus?: string;
+	boosts?: Partial<BoostsTable>;
+}
+
+interface SecondaryEffect {
+	chance?: number;
+	boosts?: Partial<BoostsTable>;
+	self?: SelfEffect;
+	status?: StatusName;
+	volatileStatus?: string;
+
+}
+interface Move extends Data {
+	kind: 'Move';
+
+	accuracy: number | true;
+	basePower: number;
+	flags: MoveFlags;
+	pp: number;
+	target: MoveTarget;
+	type: TypeName;
+
+	alwaysCrit?: boolean;
```

Comment:
Kind of wack to have all of these fields - I tried to remove the ones which were clearly for `ActiveMove`, but there's still a bunch here. What determines inclusion in `flags` vs. being a top level `boolean`, @Marty?

Reply:
We discussed this in PM but I'll reply here for everyone in some more detail.

`flags` was originally for things that were bitflagged in the games' move data, where type, power, PP, etc are also stored. When XY came out the 3DS was still an impenetrable fortress of security so some other current `flags` were assumed to exist (such as `bite`, `bullet`, `powder`), but those are actually implemented elsewhere in the game, probably as arrays of move indices in their respective effects.

There's nothing stopping other useful move categorization to be implemented here in `flags`, and in fact would be very nice to have for searching the new dex. It's just that no one has thought about what might be useful and actually switched booleans over.

Reply:
Considering `flags` is no longer the in-game bitflags, can we move to the flag system (`nomirror`, `bypassprotect` used by PSdex)

Reply:
It would fix the problem, among others, of "if you forget to flag a move in OMs, it bypasses Protect".

Reply:
But it is still the in-game bitflags, just with extra stuff. How do you think we update moves so quickly at the beginning of the gen? I have a script to pull out the bitflags and map them to the PS naming conventions. What I absolutely don't want is to putz around with questionably accurate inversions of the bitflags.

Reply:
I would at least want to invert `protect` (doesn't seem that hard for a script to invert) and give better names to the others (it's not like the names are officially set by Game Freak).

Reply:
It doesn't make any sense to flag stuff like Rain Dance as "not Protectable". I don't want to get into inverting anything at all.

Reply:
I'd still rather flag them than not.

Reply:
You can obviously flag whatever you want after the fact, which you have done in your PSdex already. e.g. https://dex.pokemonshowdown.com/tags/nonreflectable

I didn't complain about that because I can still go to https://dex.pokemonshowdown.com/tags/reflectable for the real list. If that stops being possible then I have no use for the dex and might as well use Veekun for everything.